{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b27f3378",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef87204f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_caf_output_to_snpeff_input(cohort):\n",
    "    out_dir = cohort + '/snpeff_input'\n",
    "    print(out_dir)\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)\n",
    "\n",
    "    for chrom in range(1, 23):\n",
    "        print('Chrom:', chrom)\n",
    "        cur_file = glob.glob(cohort + '/*chr' + str(chrom) + '.csv')\n",
    "\n",
    "        if len(cur_file) > 1:\n",
    "            sys.exit('More than 1 file matched for chrom ' + str(chrom))\n",
    "        else:\n",
    "            cur_file = cur_file[0]\n",
    "\n",
    "        df = pd.read_csv(cur_file)\n",
    "\n",
    "        # keep SNVs only\n",
    "        df = df.loc[ df['Variant ID'] == 'snv', ]\n",
    "\n",
    "        # Create snpEff vcf input with all unique variants\n",
    "        out_df = pd.DataFrame(index=df.index)\n",
    "        out_df.reset_index(inplace=True)\n",
    "        out_df[['chrom', 'pos', 'ref', 'alt']] = out_df['index'].str.split('-', expand=True)\n",
    "        del out_df['index']\n",
    "        out_df['aux_1'] = '.'\n",
    "        out_df['aux_2'] = '.'\n",
    "        out_df['aux_3'] = '.'\n",
    "        out_df['aux_4'] = '.'\n",
    "\n",
    "        # re-order columns\n",
    "        out_df = out_df[['chrom', 'pos', 'aux_1', 'ref', 'alt', 'aux_2', 'aux_3', 'aux_4']]\n",
    "\n",
    "        # save to output file\n",
    "\n",
    "        out_df.to_csv(out_dir + '/' + cohort + '.snpeff_input.chr' + str(chrom) + '.vcf', sep='\\t', index=False, header=False)\n",
    "\n",
    "\n",
    "    # merge output from all chroms into a single file\n",
    "    os.system('cat '+out_dir+'/'+cohort+'.snpeff_input.chr*.vcf | sort -k1 -n > ' + out_dir+'/'+cohort+'.snpeff_input.All_chr.vcf')\n",
    "    \n",
    "    print('Cohort', cohort, ' is completed.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "95aa6c01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NFE_440k/snpeff_input\n",
      "Chrom: 1\n",
      "Chrom: 2\n",
      "Chrom: 3\n",
      "Chrom: 4\n",
      "Chrom: 5\n",
      "Chrom: 6\n",
      "Chrom: 7\n",
      "Chrom: 8\n",
      "Chrom: 9\n",
      "Chrom: 10\n",
      "Chrom: 11\n",
      "Chrom: 12\n",
      "Chrom: 13\n",
      "Chrom: 14\n",
      "Chrom: 15\n",
      "Chrom: 16\n",
      "Chrom: 17\n",
      "Chrom: 18\n",
      "Chrom: 19\n",
      "Chrom: 20\n",
      "Chrom: 21\n",
      "Chrom: 22\n",
      "Cohort NFE_440k  is completed.\n"
     ]
    }
   ],
   "source": [
    "#cohorts = ['NFE_20K_WES', 'NFE_43K_WES', 'EAS_WES', 'SAS_WES', 'AFR_WES', 'ASJ_WES',\n",
    "#           'NFE_20K_WGS', 'NFE_43K_WGS', 'EAS_WGS', 'SAS_WGS', 'AFR_WGS', 'ASJ_WGS'\n",
    "#             'Maximally-diverse_WES', 'Maximally-diverse_WGS']\n",
    "\n",
    "cohorts = ['NFE_440k']\n",
    "\n",
    "for cohort in cohorts:\n",
    "    convert_caf_output_to_snpeff_input(cohort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79a950da",
   "metadata": {},
   "source": [
    "## Parse Diverse 460k cohort (single file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "674d1d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_single_caf_output_to_snpeff_input(cohort):\n",
    "    out_dir = cohort + '/snpeff_input'\n",
    "    print(out_dir)\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)\n",
    "\n",
    "    cur_file = cohort + '/compiled_full_dataset.csv'\n",
    "    df = pd.read_csv(cur_file)\n",
    "\n",
    "    # keep SNVs only\n",
    "    df = df.loc[ df['Variant Type'] == 'snv', ]\n",
    "\n",
    "    # Create snpEff vcf input with all unique variants\n",
    "    out_df = pd.DataFrame(index=df.index)\n",
    "    out_df.reset_index(inplace=True)\n",
    "    out_df[['chrom', 'pos', 'ref', 'alt']] = out_df['index'].str.split('-', expand=True)\n",
    "    del out_df['index']\n",
    "    out_df['aux_1'] = '.'\n",
    "    out_df['aux_2'] = '.'\n",
    "    out_df['aux_3'] = '.'\n",
    "    out_df['aux_4'] = '.'\n",
    "\n",
    "    # re-order columns\n",
    "    out_df = out_df[['chrom', 'pos', 'aux_1', 'ref', 'alt', 'aux_2', 'aux_3', 'aux_4']]\n",
    "\n",
    "    # save to outpu file\n",
    "\n",
    "    out_df.to_csv(out_dir + '/' + cohort + '.snpeff_input.chr' + str(chrom) + '.vcf', sep='\\t', index=False, header=False)\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "    # merge output from all chroms into a single file\n",
    "    os.system('cat '+out_dir+'/'+cohort+'.snpeff_input.chr*.vcf | sort -k1 -n > ' + out_dir+'/'+cohort+'.snpeff_input.All_chr.vcf')\n",
    "    \n",
    "    print('Cohort', cohort, ' is completed.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "944e70e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = 'Diverse_460k'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "588a5ef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Diverse_460k/snpeff_input\n"
     ]
    }
   ],
   "source": [
    "out_dir = cohort + '/snpeff_input'\n",
    "print(out_dir)\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "cur_file = cohort + '/compiled_full_dataset.csv'\n",
    "df = pd.read_csv(cur_file)\n",
    "\n",
    "# keep SNVs only\n",
    "df = df.loc[ df['Variant Type'] == 'snv', ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7ed9fbe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Function</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>afr_maf</th>\n",
       "      <th>asj_maf</th>\n",
       "      <th>eas_maf</th>\n",
       "      <th>nfe_400k_maf</th>\n",
       "      <th>sas_maf</th>\n",
       "      <th>Gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1-69496-G-A</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>0.003900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1-69511-A-G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>0.042900</td>\n",
       "      <td>0.0577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133300</td>\n",
       "      <td>0.235800</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1-69525-C-G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1-69559-G-A</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1-69581-C-G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398544</th>\n",
       "      <td>12398545</td>\n",
       "      <td>Y-20768970-G-C</td>\n",
       "      <td>RPS4Y2</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>ENSG00000280969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398545</th>\n",
       "      <td>12398546</td>\n",
       "      <td>Y-21383590-T-G</td>\n",
       "      <td>PRORY</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>ENSG00000183146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398546</th>\n",
       "      <td>12398547</td>\n",
       "      <td>Y-21383591-A-G</td>\n",
       "      <td>PRORY</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>ENSG00000183146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398547</th>\n",
       "      <td>12398548</td>\n",
       "      <td>Y-21383611-T-C</td>\n",
       "      <td>PRORY</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>ENSG00000183146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12398548</th>\n",
       "      <td>12398549</td>\n",
       "      <td>Y-21383657-A-G</td>\n",
       "      <td>PRORY</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>snv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>ENSG00000183146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11868902 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Unnamed: 0      Variant ID    Gene            Function Variant Type  \\\n",
       "0                  1     1-69496-G-A   OR4F5    missense_variant          snv   \n",
       "1                  2     1-69511-A-G   OR4F5    missense_variant          snv   \n",
       "2                  3     1-69525-C-G   OR4F5  synonymous_variant          snv   \n",
       "3                  4     1-69559-G-A   OR4F5    missense_variant          snv   \n",
       "4                  5     1-69581-C-G   OR4F5    missense_variant          snv   \n",
       "...              ...             ...     ...                 ...          ...   \n",
       "12398544    12398545  Y-20768970-G-C  RPS4Y2    missense_variant          snv   \n",
       "12398545    12398546  Y-21383590-T-G   PRORY    missense_variant          snv   \n",
       "12398546    12398547  Y-21383591-A-G   PRORY    missense_variant          snv   \n",
       "12398547    12398548  Y-21383611-T-C   PRORY    missense_variant          snv   \n",
       "12398548    12398549  Y-21383657-A-G   PRORY  synonymous_variant          snv   \n",
       "\n",
       "           afr_maf  asj_maf  eas_maf  nfe_400k_maf   sas_maf          Gene ID  \n",
       "0         0.003900      NaN      NaN      0.000007       NaN  ENSG00000186092  \n",
       "1         0.042900   0.0577      NaN      0.133300  0.235800  ENSG00000186092  \n",
       "2         0.000154      NaN      NaN           NaN       NaN  ENSG00000186092  \n",
       "3         0.000153      NaN      NaN           NaN       NaN  ENSG00000186092  \n",
       "4         0.000076      NaN      NaN           NaN       NaN  ENSG00000186092  \n",
       "...            ...      ...      ...           ...       ...              ...  \n",
       "12398544       NaN      NaN      NaN           NaN  0.000216  ENSG00000280969  \n",
       "12398545       NaN      NaN      NaN           NaN  0.000203  ENSG00000183146  \n",
       "12398546       NaN      NaN      NaN           NaN  0.000203  ENSG00000183146  \n",
       "12398547       NaN      NaN      NaN           NaN  0.000203  ENSG00000183146  \n",
       "12398548       NaN      NaN      NaN           NaN  0.000406  ENSG00000183146  \n",
       "\n",
       "[11868902 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "79d572b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cohort Diverse_460k  is completed.\n"
     ]
    }
   ],
   "source": [
    "# Create snpEff vcf input with all unique variants\n",
    "out_df = pd.DataFrame(index=df['Variant ID'])\n",
    "out_df.reset_index(inplace=True)\n",
    "out_df[['chrom', 'pos', 'ref', 'alt']] = out_df['Variant ID'].str.split('-', expand=True)\n",
    "del out_df['Variant ID']\n",
    "out_df['aux_1'] = '.'\n",
    "out_df['aux_2'] = '.'\n",
    "out_df['aux_3'] = '.'\n",
    "out_df['aux_4'] = '.'\n",
    "\n",
    "# re-order columns\n",
    "out_df = out_df[['chrom', 'pos', 'aux_1', 'ref', 'alt', 'aux_2', 'aux_3', 'aux_4']]\n",
    "\n",
    "# save to output file\n",
    "out_df.to_csv(out_dir + '/' + cohort + '.snpeff_input.All_chr.vcf', sep='\\t', index=False, header=False)\n",
    "\n",
    "print('Cohort', cohort, ' is completed.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cgrai",
   "language": "python",
   "name": "cgrai"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
